contracts/upgradeable_contracts/arbitrary_message/BasicHomeAMB.sol

```diff
@@ -0,0 +1,137 @@
+pragma solidity 0.4.24;
+
+import "../../libraries/Message.sol";
+import "./BasicAMB.sol";
+import "./MessageDelivery.sol";
+import "./MessageProcessor.sol";
+
+contract BasicHomeAMB is BasicAMB, MessageDelivery, MessageProcessor {
+    event SignedForUserRequest(address indexed signer, bytes32 messageHash);
+    event SignedForAffirmation(address indexed signer, bytes32 messageHash);
+
+    event CollectedSignatures(
+        address authorityResponsibleForRelay,
+        bytes32 messageHash,
+        uint256 NumberOfCollectedSignatures
+    );
+
+    function executeAffirmation(bytes message) external onlyValidator {
+        bytes32 hashMsg = keccak256(abi.encodePacked(message));
+        bytes32 hashSender = keccak256(abi.encodePacked(msg.sender, hashMsg));
+        // Duplicated affirmations
+        require(!affirmationsSigned(hashSender));
+        setAffirmationsSigned(hashSender, true);
+
+        uint256 signed = numAffirmationsSigned(hashMsg);
+        require(!isAlreadyProcessed(signed));
+        // the check above assumes that the case when the value could be overflew will not happen in the addition operation below
+        signed = signed + 1;
+
+        setNumAffirmationsSigned(hashMsg, signed);
+
+        emit SignedForAffirmation(msg.sender, hashMsg);
+
+        if (signed >= requiredSignatures()) {
+            setNumAffirmationsSigned(hashMsg, markAsProcessed(signed));
```

Comment:
We mark here that the message has been processed. But then we do the same in `processMessage`:
```solidity
    require(!messageProcessed(txHash));
    setMessageProcessed(txHash, true);
```
So, we spend extra gas for the same action.

Reply:
Made a refactor to avoid `processMessage` checking for this. Please take a look https://github.com/poanetwork/poa-bridge-contracts/pull/77/commits/7f8307554658a61a024de086e2cc65b40ec149b1

Reply:
should we really have two `processMessage` methods?

Reply:
Moved the logic so we have only one `processMessage` and now `BasicHomeAMB` works similar to `BasicForeignAMB` https://github.com/poanetwork/poa-bridge-contracts/pull/77/commits/a23dfba06c5eb4f48c00a855231b61160e25eeb9